infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMerger.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.environment;
+
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toCollection;
+import static org.eclipse.che.workspace.infrastructure.kubernetes.Constants.MACHINE_NAME_ANNOTATION_FMT;
+
+import io.fabric8.kubernetes.api.model.Container;
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+import io.fabric8.kubernetes.api.model.ObjectMetaBuilder;
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.api.model.PodSpec;
+import io.fabric8.kubernetes.api.model.PodTemplateSpec;
+import io.fabric8.kubernetes.api.model.apps.Deployment;
+import io.fabric8.kubernetes.api.model.apps.DeploymentBuilder;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.eclipse.che.commons.lang.NameGenerator;
+import org.eclipse.che.workspace.infrastructure.kubernetes.Names;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment.PodData;
+
+/** @author Sergii Leshchenko */
+public class PodMerger {
+
+  public Deployment merge(Map<String, Pod> pods, Map<String, Deployment> deployments) {
+    List<PodData> podData = new ArrayList<>();
+
+    pods.values().forEach(e -> podData.add(new PodData(e.getSpec(), e.getMetadata())));
+
+    deployments
+        .values()
+        .stream()
+        .map(d -> d.getSpec().getTemplate())
+        .forEach(pt -> podData.add(new PodData(pt.getSpec(), pt.getMetadata())));
+
+    return merge(createDeployment(), podData);
+  }
+
+  public Deployment merge(Deployment baseDeployment, List<PodData> pods) {
+    PodTemplateSpec basePodTemplate = baseDeployment.getSpec().getTemplate();
+    ObjectMeta baseMeta = basePodTemplate.getMetadata();
+    PodSpec baseSpec = basePodTemplate.getSpec();
+
+    Set<String> containerNames =
+        baseSpec
+            .getContainers()
+            .stream()
+            .map(Container::getName)
+            .collect(toCollection(HashSet::new));
+    Set<String> initContainerNames =
+        baseSpec
+            .getInitContainers()
+            .stream()
+            .map(Container::getName)
+            .collect(toCollection(HashSet::new));
+    for (PodData podData : pods) {
+      baseMeta.getLabels().putAll(podData.getMetadata().getLabels());
```

Comment:
In here we're potentially breaking pod selectors in services of the `KubernetesEnvironment`, am I right?

Reply:
Yes, it's why I reconfigure all services here https://github.com/eclipse/che/pull/12842/files#diff-db9fbcff9d0ee452ffca206974b7c622R144